cpp/src/aztec/honk/composer/composer_helper/permutation_helper.hpp

```diff
@@ -1,118 +1,89 @@
-#include <stdint.h>
-#include <stddef.h>
-#include <vector>
+#pragma once
+
+#include <ecc/curves/bn254/fr.hpp>
+#include <polynomials/polynomial.hpp>
 #include <proof_system/proving_key/proving_key.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <initializer_list>
+#include <cstdint>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
 namespace honk {
-// Enum values spaced in increments of 30-bits (multiples of 2 ** 30).
-enum WireType { LEFT = 0U, RIGHT = (1U << 30U), OUTPUT = (1U << 31U), FOURTH = 0xc0000000 };
 
 /**
- * @brief cycle_node represents a particular witness at a particular gate. Used to collect permutation sets
- *
+ * @brief cycle_node represents the index of a value of the circuit.
+ * It will belong to a CyclicPermutation, such that all nodes in a CyclicPermutation
+ * must have the value.
+ * The total number of constaints is always <2^32 since that is the type used to represent variables, so we can save
+ * space by using a type smaller than size_t.
  */
 struct cycle_node {
+    uint32_t wire_index;
     uint32_t gate_index;
-    WireType wire_type;
-
-    cycle_node(const uint32_t a, const WireType b)
-        : gate_index(a)
-        , wire_type(b)
-    {}
-    cycle_node(const cycle_node& other)
-        : gate_index(other.gate_index)
-        , wire_type(other.wire_type)
-    {}
-    cycle_node(cycle_node&& other)
-        : gate_index(other.gate_index)
-        , wire_type(other.wire_type)
-    {}
-    cycle_node& operator=(const cycle_node& other)
-    {
-        gate_index = other.gate_index;
-        wire_type = other.wire_type;
-        return *this;
-    }
-    bool operator==(const cycle_node& other) const
-    {
-        return ((gate_index == other.gate_index) && (wire_type == other.wire_type));
-    }
 };
-typedef std::vector<std::vector<cycle_node>> CycleCollector;
+using CyclicPermutation = std::vector<cycle_node>;
```

Comment:
All these constructors could be removed, making `cycle_node` a POD (plain-old data). Also changed the typedef to be more explicit.